### PR TITLE
[RHCLOUD-39981] Access

### PR DESCRIPTION
**Pull request by astrozzc**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-39981
## Description of Intent of Change(s)
(https://docs.google.com/document/d/1pqts4SxxXnSQq-Hcxd2RSbi95AvW2ZQtDF67ttE80xs/edit?tab=t.0)

## Local Testing
Unit tests

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

## Summary by Sourcery

Add optional handling for ungrouped hosts workspace IDs in access and role management APIs, controlled by new feature flags and centralized utility for workspace creation.

Enhancements:
- Add ADD_UNGROUPED_HOSTS_ID and REMOVE_NULL_VALUE settings to toggle inclusion or removal of null group.id filters
- Implement add_ungrouped_hosts_id in access view to append ungrouped hosts workspace ID when applicable
- Introduce get_or_create_ungrouped_workspace utility and refactor internal retrieve to reuse workspace creation logic
- Extend ResourceDefinition serializer to replace null group.id values with ungrouped hosts workspace ID based on settings

Tests:
- Add tests for null value handling in access API under various ADD_UNGROUPED_HOSTS_ID and REMOVE_NULL_VALUE configurations
- Add test to verify null group.id replacement behavior when retrieving roles with REMOVE_NULL_VALUE setting

Chores:
- Rename RelationApiDualWriteWorkspacepHandler to RelationApiDualWriteWorkspaceHandler across management and service files
- Update workspace cleanup in test teardowns to filter by workspace types instead of parent

### CODE REVIEW

**Comment by lpichler** [comment]
@astrozzc do we need to add those variables into clowder.yaml so we can control  it via app-interface ?

**Comment by lpichler** [comment]
@astrozzc 

It tried to put value with equal operation where I saw some occurrences in google sheet , like

```
{
  "name": "inventory-a3",
  "display_name": "inventory-a3",
  "description": "inventory-a3",
  "access": [
    {
      "resourceDefinitions": [
        {
          "attributeFilter": {
            "key": "group.id",
            "value": ["01968110-1cbb-71f2-9dbc-4447cc58e9a3", "0196810f-3088-7811-a109-a2856b9ee3c0"],
            "operation": "in"
          }
        }
      ],
      "permission": "inventory:groups:write"
    },
    {
      "permission": "inventory:hosts:read",
      "resourceDefinitions": [{
        "attributeFilter": {
          "key": "group.id",
          "operation": "equal",
          "value": "XXXx"
        }
      }]
    }
  ]
}
```

and I got 500 error, when I called /access:

```

Traceback (most recent call last):
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/liborpichler/Projects/insights-rbac/rbac/management/access/view.py", line 128, in get
    access_policy = self.serializer_class(queryset, many=True).data
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 795, in data
    ret = super().data
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 249, in data
    self._data = self.to_representation(self.instance)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 713, in to_representation
    return [
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 714, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 538, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 713, in to_representation
    return [
  File "/Users/liborpichler/.local/share/virtualenvs/insights-rbac-KgJzuoHm/lib/python3.9/site-packages/rest_framework/serializers.py", line 714, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/Users/liborpichler/Projects/insights-rbac/rbac/management/role/serializer.py", line 75, in to_representation
    and None in instance.attributeFilter["value"]
TypeError: 'in <string>' requires string as left operand, not NoneType
```

File: rbac/management/access/view.py [hunk 1]
```diff
@@ -161,3 +164,26 @@ def validate_and_get_param(self, params):
...
             sub_key = f"{app}&order:{ordering}"
         return sub_key, ordering
+
+    def add_ungrouped_hosts_id(self, response, tenant):
+        """Add ungrouped hosts id to the data."""
+        if not settings.ADD_UNGROUPED_HOSTS_ID:
+            return
+        ungrouped_hosts_id = None
+        queried = False
+        for access in response.data["data"]:
```
Comment (lpichler) [inline rbac/management/access/view.py:174]: can we just scope it for `inventory:hosts:XXX` permissions ? I think nulls could only for `inventory:host` permissions.

Comment (astrozzc) [reply rbac/management/access/view.py:174]: Sure, updated

**Review by lpichler — APPROVED** [review]
